File: 018-树的子结构/hassubtree.py
# -*- coding:utf-8 -*-
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
class Solution:
    def HasSubtree(self, pRoot1, pRoot2):
        # write code here
        if pRoot2 is None:
            return False
        if pRoot1 is None and pRoot2 is not None:
            return False
        tag=False
        if pRoot2.val==pRoot1.val:
            tag=self.issubtree(pRoot1,pRoot2)
        if not tag:
            tag=self.HasSubtree(pRoot1.left,pRoot2)
            if not tag:
                tag=self.HasSubtree(pRoot1.right,pRoot2)
        return tag
    def issubtree(self,pRoot1, pRoot2):
        if pRoot2 is None:
            return True
        if pRoot2 is not None and pRoot1 is None:
            return False
        if pRoot2.val==pRoot1.val:
            return self.issubtree(pRoot1.left,pRoot2.left) and self.issubtree(pRoot1.right,pRoot2.right)
        else:
            return False

File: 018-树的子结构/hassubtree.cpp
/*
struct TreeNode {
	int val;
	struct TreeNode *left;
	struct TreeNode *right;
	TreeNode(int x) :
			val(x), left(NULL), right(NULL) {
	}
};*/
class Solution {
public:
    bool HasSubtree(TreeNode* pRoot1, TreeNode* pRoot2)
    {
        if (pRoot2==nullptr){
            return false;
        }
        if (pRoot2!=nullptr && pRoot1==nullptr){
            return false;
        }
        bool tag=false;
        if(pRoot1->val==pRoot2->val){
            tag=issubtree(pRoot1,pRoot2);
        }
        if(!tag){
            tag=HasSubtree(pRoot1->left,pRoot2);
            if(!tag){
                tag=HasSubtree(pRoot1->right,pRoot2);
            }
        }
        return tag;
    }
    bool issubtree(TreeNode* pRoot1, TreeNode* pRoot2){
        if (pRoot2==nullptr){
            return true;
        }
        if (pRoot2!=nullptr && pRoot1==nullptr){
            return false;
        }
        if(pRoot1->val==pRoot2->val){
            return issubtree(pRoot1->left,pRoot2->left) && issubtree(pRoot1->right,pRoot2->right);
        }
        return false;
    }
};
